Fixed #34309 -- Doc'd how to fully delete an app. (#16904)
django/django: docs/howto/delete-app.txt

==================================
How to delete a Django application
==================================

Django provides the ability to group sets of features into Python packages
called :doc:`applications</ref/applications/>`. However, when requirements
change, apps may become obsolete or unnecessary. The following guide will help
you delete an application safely.

1. Remove all references to the app (i.e imports, foreign keys).

> 2. Remove all models from your ``models.py`` file.

Reviewer: I think saying "your models.py" could be confusing since a project may have many models.py that are "yours" for a developer.

```suggestion
2. Remove all models from the corresponding ``models.py`` file.
```

Author: Applied!

> 3. Generate your migrations running :djadmin:`makemigrations`.

Reviewer: ```suggestion
3. Create relevant migrations by running :djadmin:`makemigrations` (this step will generate a migration that removes all tables for the removed models).
```

Author: Applied! I extended it a bit and wrote `[..] and any other migration in case there was a relationship` in case step 1 there were foreign keys deleted and generate other migrations

4. :ref:`Squash<migration-squashing>` out references to the app in other apps' migrations.

> 5. Apply migrations locally and deploy your Django application too.

Reviewer: ```suggestion
5. Apply migrations locally, runs tests and verify the correctness of your project.
6. Deploy/release your updated Django project.
```

Author: Applied!

==================================
How to delete a Django application
==================================

Django provides the ability to group sets of features into Python packages

> called :doc:`applications</ref/applications/>`. However, when requirements

Reviewer 2: `However` sounds unnecessary to me.

> change, apps may become obsolete or unnecessary. The following guide will help

Reviewer 2: Maybe:
```suggestion
change, apps may become obsolete or unnecessary. The following steps will help
```

you delete an application safely.

> 1. Remove all references to the app (i.e imports, foreign keys).

Reviewer 2: You can use `#.` for numbered list:
```suggestion
#. Remove all references to the app (i.e imports, foreign keys).
```

2. Remove all models from the corresponding ``models.py`` file.

3. Create relevant migrations by running :djadmin:`makemigrations` (this step
   will generate a migration that removes all tables for the removed models
   and any other migration in case there was a relationship).

4. :ref:`Squash<migration-squashing>` out references to the app in other apps' migrations.

> 5. Apply migrations locally, runs tests and verify the correctness of your project.

Reviewer 2: Wrap docs at 79 chars.

> 5. Apply migrations locally, runs tests and verify the correctness of your project.

Reviewer 2: ```suggestion
5. Apply migrations locally, runs tests, and verify the correctness of your project.
```

==================================
How to delete a Django application
==================================

Django provides the ability to group sets of features into Python packages
called :doc:`applications</ref/applications/>`. When requirements
change, apps may become obsolete or unnecessary. The following steps will help
you delete an application safely.

#. Remove all references to the app (i.e imports, foreign keys).

#. Remove all models from the corresponding ``models.py`` file.

#. Create relevant migrations by running :djadmin:`makemigrations` (this step
   will generate a migration that removes all tables for the removed models

>    and any other migration in case there was a relationship).

Reviewer: ```suggestion
this step generates a migration that deletes tables for the removed models, and any other required migration for updating relationships connected to those models.
```

#. :ref:`Squash<migration-squashing>` out references to the app in other apps'
   migrations.

#. Apply migrations locally, runs tests, and verify the correctness of your project.

> 

Author: I didn't see it when you suggested it and didn't add it! I'll do! Sorry!

==========================
How to delete a Django app

> ==========================

Reviewer 3: Seeing as this is a "how to" we need some sort of introduction whether it's a single sentence or a paragraph before diving straight into step by step instructions

> ==========================

Author: Applied!

Let me know if you have something in mind. I'll share it on the Discord to get more ideas and ask for help with the english.

> Before deleting an application from your project you should consider that the
> app A that you are about to delete may have migrations on which app B depends
> and in case you want to remove B in the future you may also encounter an error
> like this when you want to revert migrations or running a :djadmin:`squashmigrations`:
> 
> .. code-block:: shell
> 
>     django.db.migrations.exceptions.NodeNotFoundError: Migration
>     <app_B>.XXXX_migration_name dependencies reference
>     nonexistent parent node ('<app_A>', 'XXXX_migration_name')
> 
> Once this is said, the steps to follow are:

Reviewer 3: ```suggestion
```
I find this preamble to be confusing to follow. I reckon it's definitely enough to just add a step to say "squash out references to the app in other apps' migrations".

Author: Agree and deleted!

1. Remove all references to the app (i.e imports, foreign keys).

2. Remove all models from your ``models.py`` file.

3. Generate your migrations running :djadmin:`makemigrations`.

> 4. (Optional) In case other application (app B) has dependencies to the app to

Reviewer 3: - This shouldn't be optional, references to a deleted app in migrations shouldn't be allowed imho 😊
- Example naming (app a, app b) is unnecessary
- Shouldn't this step be first? 🤔 Otherwise squashing will break?

Author: > - This shouldn't be optional, references to a deleted app in migrations shouldn't be allowed imho 😊
> - Example naming (app a, app b) is unnecessary

Agree and applied!

> - Shouldn't this step be first? 🤔 Otherwise squashing will break?

I think both ways are the same. Because if I delete the app first it will generate 2 migrations: one on the app side and other for the app where de dependecy is and the squash will result in the same.

Author: Thinking on it, maybe putting it in the first place would give it more importance!

==========================
How to delete a Django app
==========================

To delete an app from you project you need to follow the next steps:

1. Remove all references to the app (i.e imports).

> 2. Run ``python manage.py migrate <app_label> zero``.

Author: Need to clarify **local** and **production** enviroment?

3. Remove your app from :setting:`INSTALLED_APPS` and delete the directory.

4. Run :djadmin:`remove_stale_contenttypes` command passing
   ``--include-stale-apps``.

.. warning::

    It could happen in the future you want delete other app and it has
    dependencies from the app you previously removed and get an error like:

    .. code-block:: shell

        django.db.migrations.exceptions.NodeNotFoundError: Migration
        <app_label_to_delete>.XXXX_migration_example dependencies reference

>         nonexistent parent node ('<app_label_deleted>', 'XXXX_migration_example')

Author: I created a repo to exemplify this case: https://github.com/mgaligniana/ticket_34309/commits/main

1. Clone the project
2. `python -m venv env`
3. `source env/bin/activate`
4. `pip install django`
5. Try do the first step to delete Books app doing `python manage.py migrate books zero`

    In that case is recommended replace step 2 by:
    
    2. Delete all models inside your ``<app_label>/models.py``, create the
    respective migration: ``python manage.py makemigrations <app_label>``,

>     deploy changes and run ``python manage.py migrate`` to impact changes to

Author: Need to clarify **local** and **production** enviroment?
